Project-Hotpot/Project-Hotpot/Algorithm/SongAlgorithm.swift

```diff
@@ -24,28 +26,36 @@ class SongAlgorithm{
                       let randomSong = items.randomElement(),
                       let randomURI = randomSong?["uri"]
                 else{
-                    NSLog("failed parse")
+                    NSLog("failed parsing random song dictionary response")
                     return completion(nil, error)
                 }
                 return completion(randomURI, error)
             }
         }
     }
     
+    func getRandomGenre() -> String? {
+        let genre = api_instance.genreSeedArray.randomElement()
+        return genre
+    }
+    
     //TODO: SPLIT INTO SMALLER HELPER FUNCTIONS FOR SAKE OF REUSE. CONSIDER MOVING PARTS OF THIS INTO API MANAGER
     func fetchSong(completion: @escaping ([String: Any]?, Error?) -> Void) {
-        let api_instance = SpotifyManager.shared()
-        //    I in the future will switch out this endpoint to access different search features
-        // currently just gets queries for 50 songs from a search of 'a'
-        let url = URL(string: "https://api.spotify.com/v1/search?q=a&type=track&limit=50")!
+        // currently queries 50 songs from a random genre, with a random offset
+        
+        let randomOffset = String(Int.random(in: 1..<800))
+        NSLog("randomOffset: \(randomOffset)")
+        
+        guard let genre = getRandomGenre(),
```

Comment:
There is a fair bit of code that is lower level network logic and not at all part of your Algorithm (This class). This is something you could offload to the SpotifyManager. Consider creating a method fetchSongsByGenreAndOffset(...). Generally you would want all your spotify code encapsulated there exposing easy methods for other parts of the system to call into.

Whenever you add logic or method always ask yourself whether it fits there. A funny but effective way to think about it is to think you are "SongAlgorithm", do you need to know how you are getting data or where from? There could be multiple levels of this, So you could ask yourself (as the algorithm class):
 - Should i know about Spotify at all?
 - Should i know about how the data gets to me?
 - Should i just be talking to a Data Fetcher which could have multiple data sources like spotify, soundcloud etc. as "Data Providers" encapsulated within the fetcher?

Reply:
I second the "impersonate your objects" method for thinking through what should belong where! This is very effective for me as well.